cedar-policy-core/src/ast/entity.rs

```diff
@@ -28,38 +28,71 @@ use serde::{Deserialize, Serialize};
 use serde_with::{serde_as, TryFromInto};
 use smol_str::SmolStr;
 use std::collections::{BTreeMap, HashMap, HashSet};
+use std::str::FromStr;
 use thiserror::Error;
 
-/// We support two types of entities. The first is a nominal type (e.g., User, Action)
-/// and the second is an unspecified type, which is used (internally) to represent cases
-/// where the input request does not provide a principal, action, and/or resource.
+/// Entity Type Names are just Name's, but we have some operations on them specific to Entity Types.
 #[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone, Hash, PartialOrd, Ord)]
 #[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
-pub enum EntityType {
-    /// Concrete nominal type
-    Specified(Name),
-    /// Unspecified
-    Unspecified,
-}
+#[serde(transparent)]
+pub struct EntityType(Name);
 
 impl EntityType {
     /// Is this an Action entity type
     pub fn is_action(&self) -> bool {
-        match self {
-            Self::Specified(name) => name.basename() == &Id::new_unchecked("Action"),
-            Self::Unspecified => false,
-        }
+        self.0.basename() == &Id::new_unchecked("Action")
+    }
+
+    /// The name of this entity type
+    pub fn name(&self) -> &Name {
+        &self.0
+    }
+
+    /// Construct a new [`EntityType`]
+    pub fn new(name: Name) -> Self {
+        Self(name)
+    }
+
+    /// Get the source location for this Entity Type
+    pub fn loc(&self) -> Option<&Loc> {
+        self.0.loc()
+    }
+
+    /// Prefix the name with a optional namespace
+    /// When the name is not an `Id`, it doesn't make sense to prefix any
+    /// namespace and hence this method returns a copy of `self`
+    /// When the name is an `Id`, prefix it with the optional namespace
+    /// e.g., prefix `A::B`` with `Some(C)` or `None` produces `A::B`
+    /// prefix `A` with `Some(B::C)` yields `B::C::A`
```

Comment:
Rather than duplicating this (somewhat confusing) comment, just say something like
```
/// Call [`Name::prefix_namespace_if_unqualified`] on the underlying name
```